linera-base/src/crypto/secp256k1.rs

```diff
@@ -4,39 +4,181 @@
+        }
+    }
+}
+
+impl FromStr for Secp256k1PublicKey {
+    type Err = CryptoError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let pk = secp256k1::PublicKey::from_str(s)
+            .map_err(|_| CryptoError::IncorrectPublicKeySize(0))?;
+        Ok(Secp256k1PublicKey(pk))
+    }
+}
+
+impl From<[u8; 33]> for Secp256k1PublicKey {
+    fn from(value: [u8; 33]) -> Self {
+        let pk = secp256k1::PublicKey::from_slice(&value).expect("Invalid public key");
+        Secp256k1PublicKey(pk)
+    }
+}
+
+impl TryFrom<&[u8]> for Secp256k1PublicKey {
+    type Error = CryptoError;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        let pk = secp256k1::PublicKey::from_slice(value)
+            .map_err(|_| CryptoError::IncorrectPublicKeySize(value.len()))?;
+        Ok(Secp256k1PublicKey(pk))
+    }
+}
+
+impl fmt::Display for Secp256k1PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let s = hex::encode(self.0.serialize());
+        write!(f, "{}", s)
+    }
+}
+
+impl fmt::Debug for Secp256k1PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", hex::encode(&self.0.serialize()[0..9]))
+    }
+}
+
+impl Secp256k1KeyPair {
+    #[cfg(all(with_getrandom, with_testing))]
+    /// Generates a new key-pair.
+    pub fn generate() -> Self {
+        let mut rng = rand::rngs::OsRng;
+        Self::generate_from(&mut rng)
+    }
+
+    #[cfg(with_getrandom)]
+    /// Generates a new key-pair from the given RNG. Use with care.
+    pub fn generate_from<R: super::CryptoRng>(rng: &mut R) -> Self {
+        let (sk, pk) = SECP256K1.generate_keypair(rng);
+        Secp256k1KeyPair {
+            secret_key: Secp256k1SecretKey(sk),
+            public_key: Secp256k1PublicKey(pk),
+        }
+    }
+}
+
+impl Secp256k1SecretKey {
+    /// Returns a public key for the given secret key.
+    pub fn to_public(&self) -> Secp256k1PublicKey {
+        Secp256k1PublicKey(self.0.public_key(&SECP256K1))
+    }
+}
+
+impl Secp256k1PublicKey {
+    /// Returns a public key for the given secret key.
+    #[allow(dead_code)]
+    fn from_secret_key(secret: &Secp256k1SecretKey) -> Self {
+        secret.to_public()
+    }
+}
+
 impl Secp256k1Signature {
     /// Computes a secp256k1 signature for `value` using the given `secret`.
     /// It first serializes the `T` type and then creates the `CryptoHash` from the serialized bytes.
-    pub fn new<'de, T>(value: &T, secret: &secp256k1::SecretKey) -> Self
+    pub fn new<'de, T>(value: &T, secret: &Secp256k1SecretKey) -> Self
     where
         T: BcsSignable<'de>,
     {
-        let secp = secp256k1::Secp256k1::new();
+        let secp = secp256k1::Secp256k1::signing_only();
         let message = Message::from_digest(CryptoHash::new(value).as_bytes().0);
-        let signature = secp.sign_ecdsa(&message, secret);
+        let signature = secp.sign_ecdsa(&message, &secret.0);
```

Comment:
Why not use `SecretKey::sign_ecdsa` directly? You are using `SECP256K1` above, so I assume we require the `global-context` feature anyway?

Reply:
That's "our" SECP256K1 and not from the crate. That feature (`global-context`) is mutually exclusive with some other features that we use.

Reply:
Ah, right, makes sense!